drivers/i2c/i2c_max32_rtio.c

```diff
@@ -0,0 +1,454 @@
+	struct k_sem xfer;
+	int err;
+#endif
+};
+
+static int max32_configure(const struct device *dev,
+				uint32_t dev_cfg)
+{
+	struct i2c_rtio *const ctx = ((struct max32_i2c_data *)
+		dev->data)->ctx;
+
+	return i2c_rtio_configure(ctx, dev_cfg);
+}
+
+static int max32_do_configure(const struct device *dev, uint32_t dev_cfg)
+{
+	int ret = 0;
+	const struct max32_i2c_config *const cfg = dev->config;
+	mxc_i2c_regs_t *i2c = cfg->regs;
+
+	switch (I2C_SPEED_GET(dev_cfg)) {
+	case I2C_SPEED_STANDARD: /** I2C Standard Speed: 100 kHz */
+		ret = MXC_I2C_SetFrequency(i2c, MXC_I2C_STD_MODE);
+		break;
+
+	case I2C_SPEED_FAST: /** I2C Fast Speed: 400 kHz */
+		ret = MXC_I2C_SetFrequency(i2c, MXC_I2C_FAST_SPEED);
+		break;
+
+#if defined(MXC_I2C_FASTPLUS_SPEED)
+	case I2C_SPEED_FAST_PLUS: /** I2C Fast Plus Speed: 1 MHz */
+		ret = MXC_I2C_SetFrequency(i2c, MXC_I2C_FASTPLUS_SPEED);
+		break;
+#endif
+
+#if defined(MXC_I2C_HIGH_SPEED)
+	case I2C_SPEED_HIGH: /** I2C High Speed: 3.4 MHz */
+		ret = MXC_I2C_SetFrequency(i2c, MXC_I2C_HIGH_SPEED);
+		break;
+#endif
+
+	default:
+		/* Speed not supported */
+		return -ENOTSUP;
+	}
+
+	return ret;
+}
+
+static void max32_complete(const struct device *dev, int status);
+
+static int max32_msg_start(const struct device *dev, uint8_t flags,
+				 uint8_t *buf, size_t buf_len, uint16_t i2c_addr)
+{
+	int ret = 0;
+	const struct max32_i2c_config *const cfg = dev->config;
+	struct max32_i2c_data *data = dev->data;
+	mxc_i2c_regs_t *i2c = cfg->regs;
+	mxc_i2c_req_t *req = &data->req;
+	uint8_t target_rw;
+
+	req->i2c = i2c;
+	req->addr = i2c_addr;
+
+	k_sem_take(&data->lock, K_FOREVER);
+
+	if (data->second_msg_flag == 0) {
+		MXC_I2C_ClearRXFIFO(i2c);
+		MXC_I2C_ClearTXFIFO(i2c);
+		MXC_I2C_SetRXThreshold(i2c, 1);
+
+		/* First message should always begin with a START condition */
+		flags |= I2C_MSG_RESTART;
+	}
+
+	if (flags & I2C_MSG_READ) {
+		req->rx_buf = (unsigned char *)buf;
+		req->rx_len = buf_len;
+		req->tx_buf = NULL;
+		req->tx_len = 0;
+		target_rw = (i2c_addr << 1) | 0x1;
+	} else {
+		req->tx_buf = (unsigned char *)buf;
+		req->tx_len = buf_len;
+		req->rx_buf = NULL;
+		req->rx_len = 0;
+		target_rw = (i2c_addr << 1) & ~0x1;
+	}
+	data->flags = flags;
+	data->readb = 0;
+	data->written = 0;
+	data->err = 0;
+
+	MXC_I2C_ClearFlags(i2c, ADI_MAX32_I2C_INT_FL0_MASK, ADI_MAX32_I2C_INT_FL1_MASK);
+	MXC_I2C_EnableInt(i2c, ADI_MAX32_I2C_INT_EN0_ERR, 0);
+	Wrap_MXC_I2C_SetRxCount(i2c, req->rx_len);
+	if ((data->flags & I2C_MSG_RESTART)) {
+		MXC_I2C_EnableInt(i2c, ADI_MAX32_I2C_INT_EN0_ADDR_ACK, 0);
+		MXC_I2C_Start(i2c);
+		Wrap_MXC_I2C_WaitForRestart(i2c);
+		MXC_I2C_WriteTXFIFO(i2c, &target_rw, 1);
+	} else {
+		if (req->tx_len) {
+			data->written = MXC_I2C_WriteTXFIFO(i2c, req->tx_buf, 1);
+			MXC_I2C_EnableInt(i2c, ADI_MAX32_I2C_INT_EN0_TX_THD, 0);
+		} else {
+			MXC_I2C_EnableInt(i2c, ADI_MAX32_I2C_INT_EN0_RX_THD, 0);
+		}
+	}
+
+	ret = k_sem_take(&data->xfer, K_FOREVER);
```

Comment:
This isn't quite right. _start should be non-blocking and callable from an ISR (e.g. a GPIO interrupt triggering a sensor read). k_sem usage here will prevent that.

Starting a transfer should start an event (interrupt) driven state machine that works through the requested i/o queue. Take a closer look at the i2c_twi_rtio.c version of this to see how that work with i2c_rtio.

Reply:
Hi @teburd,
I used i2c_max32.c regular driver as base since it already used I2C ISR and wanted to use same approach. Did I understood correctly the idea is to have a state machine and exclude semaphores?

Reply:
Hi @teburd,
Just a reminder for this thread, to confirm if I understood you correctly?

Reply:
Please take a look at how the nrfx twi driver is written to see what I mean, there's no semaphores involved in the submit path at all

https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/i2c/i2c_nrfx_twi_rtio.c#L145
https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/i2c/i2c_nrfx_twi_rtio.c#L37

Reply:
> Hi @teburd, I used i2c_max32.c regular driver as base since it already used I2C ISR and wanted to use same approach. Did I understood correctly the idea is to have a state machine and exclude semaphores?

yes I think you understood correctly, but the best answer is to look at the existing drivers that work the way I'd expect. i2c_nrfx_twi_rtio.c is a good place to look.

Reply:
Hi @teburd,
I tested the driver and it is working properly without semaphores (checked my logic analyzer waves for sensor data), with removed semaphores I wanted to ask is it possible to create race/access problem to particular i2c bus, here is what i'm thinking, currently in CMake its either i2c regular or i2c rtio driver but if have multiple slaves on the same i2c bus is there a scenario where one sensor is reading data and second sensor have an interrupt wouldn't that create some race problem. Also is this a not likely scenario and is it worth considering it assuming they both use rtio?

Reply:
Hi @teburd,
I wanted to ask maybe if something like this was already tested, also wanted to ping you on this thread.

Reply:
Hi @teburd ,
I've pushed my changes so that you can see how it is now without any blocking semaphores.
If I understood correctly the rtio queue implementation should take care of multiple slaves.
Here is my proposition of how tests can be implemented:
1. For example two sensors both using rtio i2c drivers on the same bus - we expect this to work and would like to test it in the future.
2. In this one if we use two sensors but one is regular i2c and the other is i2c rtio driver, here the only thing is to try and create timing situations where one will try to access bus while the other is working something (i.e. reading data)
Could you please take a look at the code first and this tests can be draft for future if we think/want to create them later.

Reply:
The tests can come later, we have testing to ensure the interface works correctly but not the sort of fuzz testing that might suss out a race if there is one, that might require something more like multiple fast i2c peripherals on the bus and semi-randomly triggering transactions with interrupts or SMP cores. Non-trivial in my view.